### DISCUSSION TITLE
Is there a decent way to produce nested raw string literals?

### DISCUSSION THREAD

**Discussion by taori** [dotnet/csharplang#6524] [General]
Currently working on a codegenerator which made me think of testing raw string literals.

I have a scenario where i effectively do this:
var sb = new StringBuilder();

```cs
sb.Append(
$$"""
<tab><tab>if(....)
<tab><tab><tab>someExecution()

<tab><tab>{{GenerateSomeMoreCodeWithParameters(properties)}} <!------ nested generation
"""
);
```

```cs
void GenerateSomeMoreCodeWithParameters(IEnumerable<PropertyDeclarationSyntax> properties)
{
	var sb = new StringBuilder();
	foreach (var property in properties)
	{
		var names = (real: property.Identifier.Text, camel: property.Identifier.Text.ToCamelCase());

		sb.AppendLine(
		$$"""
		if (_handler.Get{{names.real}}Completions() is var {{names.camel}}Completions && {{names.camel}}Completions.Length > 0)
		""");
		sb.AppendLine(
		$$"""			
		<TAB>{{names.real}}.AddCompletions({{names.camel}}Completions);
		"""
		);
	}
}
```

This will generate code like this:

<TAB><TAB>if(....)
<TAB><TAB><TAB>someExecution()

<TAB><TAB>if (_handler.GetSomePropertyCompletions() is var somePropertyCompletions && somePropertyCompletions.Length > 0)
<TAB>SomeProperty.AddCompletions(somePropertyCompletions);

I suppose during language design the issues with nested raw string literals came up but exceeded the scope?

Is there a recommended way to deal with such scenarios? There are not many examples about this i could find yet.

**Comment by 333fred** [comment]
I'm not seeing where you want to nest the literals in this example?

**Reply by taori** [reply]
{{GenerateSomeMoreCodeWithParameters(properties)}} this bit is the nested literal bit

**Reply by jnm2** [reply]
A nested literal would be `{{"""`...

`{{MethodCall(foo)}}` is not a nested literal, since `MethodCall(foo)` is method invocation syntax not string literal syntax.

**Reply by jnm2** [reply]
Here is an example of a nested literal ([sharplab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1TATgAoASAIirwAI6aAxASzAGdgaAbJqAU1voBvKhQH16AOV7tecLk2C8wAQ05jxNEQF91AGWXt5fdSICUAbjxA===)):

```cs
Console.WriteLine($"""
    First line
    {"""
        Nested literal
        """}
    Last line
    """);
```

Without knowing how you would like to make use of nested literals, it's hard to say much more. Do you want to use them to simplify your generation code? Do you want to use them in the output of your generation code?

**Reply by jnm2** [reply]
My best guess at what you're asking (and this might be way off) is if raw string literals can enable you to return `void` instead of `string` from GenerateSomeMoreCodeWithParameters, and use a single StringBuilder the whole way through. I don't see how this is related to raw string literals any more than the interpolated string literals that C# has always had, though.

**Reply by taori** [reply]
your snipet is great to reproduce my issue actually

[this](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1TATgAoASAIirwAI6aAxASzAGdgaAbJqAU1vqCA3lQoDBdAIbiJggEYzZsgMaL6ogL4yAMpPZce/XINEBKANx4gA==) points out my problem best i think. i would expect b+c to be indented more than they are.

**Reply by CyrusNajmabadi** [reply]
```c#
using System;

Console.WriteLine($"""
    First line
        {"""
        a
            b
                c
        """}
    Last line
    """);
```

This is working as intended.  Let's start with the inner raw literal.  the token for that is:

```c#
"""
        a
            b
                c
        """
```

Given the end quotes, the actual content value of that string is:

```
'a
    b
        c'
```

(where the quotes designature the start/end of hte content).

Now, that content is embedded into the following:

```
FirstLine
    $$
LastLine
```

(where the dollars indicate where tehy will be inserted).  So if you insert:

```
'a
    b
        c'
```

into that, you get:

```
FirstLine
    a
    b
        c
LastLine
```

Which is exactly as intended here.

**Reply by CyrusNajmabadi** [reply]
Here's another way of thinking about it.  Instead of seeing it as a nested literal that should be impacted by the outer, imagine you just wrote this:

```c#
using System;

Console.WriteLine($"""
    First line
        {ComputeInner()}
    Last line
    """);

string ComputeInner()
{
    return """
        a
            b
                c
        """;
}
```

The interpolations in C# never mutate their contents.  So whatever their contents are, however they were obtained, is what will be injected in the final string.  In this case, regardless of whether you got that inner string from a literal, or as the result of a method call, we take that content and inject it unmolested.

**Reply by taori** [reply]
Yeah i think it is perfecly reasonable to leave the implementation as is because it produces technically reliable results. I think the easiest way to move forward, for people who experience the same issue, is to point them to your snippet right above or [this](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1TATgAoASAIirwAI6aAxASzAGdgaAbJqAU1voBvKhQH16AQzHiZAI2kzFNAMbSRAX2kAZCey49+ucSICUAbjxA).

Resetting the indentation prior to inserting a nested literal produces perfectly understandable indentation cases. Thanks for your guidance with this.

**Reply by CyrusNajmabadi** [reply]
You're welcome :-)

**Comment by CyrusNajmabadi** [comment]
Nested literals should work fine. I'm not seeing what the problem is.

**Reply by taori** [reply]
technically they do work fine and they render the way i would expect them to, but for this use case they behave oddly if the nested string literal has a multiple lines. basically i would expect them to "reset" the left indent further in depending on where a subliteral is injected. not sure if it's clear what i mean

**Reply by CyrusNajmabadi** [reply]
> basically i would expect them to "reset" the left indent further in depending on where a subliteral is injected. not sure if it's clear what i mean

This is not how they work :).  Each raw string literal is independent, and its indentation is controlled entirely and only by it's end `"""` line.  Thanks!
